File: cp-trials/src/main/java/in/shabhushan/cp_trials/math/logic/TwoSum.java
package in.shabhushan.cp_trials.math.logic;

import java.util.HashMap;

public class TwoSum {
    public static int[] twoSum(int[] numbers, int target) {
        var numToIndex = new HashMap<Integer, Integer>();

        for (int index = 0; index < numbers.length; index++) {
            numToIndex.put(numbers[index], index);
        }

        for (int index = 0; index < numbers.length; index++) {
            int secondNum = target - numbers[index];

            if (numToIndex.containsKey(secondNum)) {
                return new int[]{
                        index,
                        numToIndex.get(secondNum)
                };
            }
        }

        return new int[]{};
    }
}

File: cp-trials/src/main/kotlin/in/shabhushan/cp_trials/logic/TwoSum.kt
package `in`.shabhushan.cp_trials.logic

object TwoSum {
  fun twoSum2(
    nums: IntArray,
    target: Int
  ): IntArray {
    val map = nums.mapIndexed { index, element ->
      element to index
    }.toMap()

    val keys = map.keys

    nums.forEachIndexed { index, element ->
      val otherNumber = target - element

      if (otherNumber in keys && map[otherNumber] != index)
        return intArrayOf(index, map[otherNumber] ?: error("err-other-number-not-found"))
    }

    throw IllegalArgumentException("err-no-two-sum-solution")
  }

  fun twoSum(
    nums: IntArray,
    target: Int
  ): IntArray {
    val map = mutableMapOf<Int, Int>()

    nums.forEachIndexed { index, element ->
      val otherNumber = target - element

      if (otherNumber in map.keys)
        return intArrayOf(map[otherNumber] ?: error("err-other-number-not-found"), index)

      map[element] = index
    }

    throw IllegalArgumentException("err-no-two-sum-solution")
  }
}
